python/paddle/fluid/contrib/int8_inference/utility.py

```diff
@@ -0,0 +1,700 @@
+
+    def __generate_output_program(self):
+        for i in self.program.list_vars():
+            if not i.persistable and i.name in self._weights_var_name + self._conv_input_var_name + self._conv_output_var_name + self._residual_input_var_name:
+                i.persistable = True
+                self._persistable_vars.append(i.name)
+
+        self._output_program = self.program.clone()
+
+    def __save_scale(self):
+        '''
+        Update the convolution scale information.
+        '''
+        func = self.__get_optimal_scaling_factor if self.algo == 'KL' else np.max
+
+        for i in self._conv_op_index[1:]:
+            weights_var_name = self.program.current_block().ops[i].input(
+                'Filter')[0]
+            input_var_name = self.program.current_block().ops[i].input('Input')[
+                0]
+            output_var_name = self.program.current_block().ops[i].output(
+                'Output')[0]
+            self._output_program.current_block().ops[i]._set_attr(
+                "Scale_weights", self._var_max_value_map[weights_var_name])
+            self._output_program.current_block().ops[i]._set_attr(
+                "Scale_in", self._var_max_range[input_var_name] /
+                func(self._var_max_value_map[input_var_name]))
+
+            self._output_program.current_block().ops[i]._set_attr(
+                "Scale_out", self._var_max_range[output_var_name] /
+                func(self._var_max_value_map[output_var_name]))
+            if self._output_program.current_block().ops[i].desc.input(
+                    "ResidualData"):
+                residual_var_name = self._output_program.current_block().ops[
+                    i].desc.input("ResidualData")[0]
+                self._output_program.current_block().ops[i]._set_attr(
+                    "Scale_in_eltwise", self._var_max_range[residual_var_name] /
+                    func(self._var_max_value_map[residual_var_name]))
+
+    def __sampling(self, sampling_data):
+        '''
+        Sampling the variables data range.
+        '''
+        for i in self.program.list_vars():
+            if i.name not in self._weights_var_name + self._conv_input_var_name + self._conv_output_var_name + self._residual_input_var_name:
+                continue
+
+            if i.name in self._weights_var_name:
+                max_value = []
+                data = sampling_data[i.name][0]
+                for j in range(data.shape[0]):
+                    var_value = float(np.max(np.abs(data[j])))
+                    if not self._is_close(var_value, 0.0):
+                        max_value.append(Calibrator.s8_max / var_value)
+                    else:
+                        max_value.append(0.0)
+                max_range = Calibrator.s8_max
+            else:
+                if i.name in self._conv_output_var_name:
+                    cur_op = self.program.current_block().ops[
+                        self._conv_output_var_name.index(i.name) + 2]
```

Comment:
Yes, it's bug because it should be indexed again as a parameter of self._conv_op_index.
Anyway, i abandoned this implementation as it may raise confuse. Please see my changes later.